### PR TITLE
cacher: don't accept requests if stopped

### PR DESCRIPTION
**Pull request by aojea**
The cacher blocks until it is ready to answer request, however, the
ready variable doesn't differentiate if the cacher was stopped.

Before blocking waiting for the cacher to be ready, check if it
was stopped and return an error notifying the client.

/kind bug

```release-note
kube-apiserver: Get, GetList and Watch requests that should be served by the apiserver cacher during shutdown will be rejected to avoid a deadlock situation leaving requests hanging.
```

### CODE REVIEW

**Comment by aojea** [comment]
I set the check before the ready() condition, but I don't know if it should short circuit immediately , not familiar with this code :shrug:

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 1]
```diff
@@ -462,6 +462,9 @@ func (c *Cacher) Watch(ctx context.Context, key string, opts storage.ListOptions
 		return nil, err
 	}
 
+	if c.stopped {
+		return nil, storage.NewInternalError("cacher stopped")
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:466]: This is still racy, right?

Can we change that to something like:

1. use cacher.stopCh that we already have
2. pass that down to cacher.ready (via newReady)
3. newReady() starting a goroutine that will wait on that channel and on its closure switch ready to "stopped"
4. switch ready.wait() to instead of returning bool, additionally to be able to return an error (and the only error that it will be returning for now will be "stopped")
?

The above seems not visibly more complicated and non-racy.

**Comment by aojea** [comment]
this actually reduces the number of goroutines leaked by 
```
s := apitesting.StartTestServerOrDie(t, nil, nil, etcdConfig)
	s.TearDownFn()
```

from ~200 to 60

grep expected test.log 
    teardown_test.go:63: expected 4 or fewer, got 59

https://github.com/kubernetes/kubernetes/pull/108060#issuecomment-1053672987

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go (hunk 1 above)
Comment (aojea) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:466]: yeah, that sounds more correct , I was afraid of changing a conditional variable for that ...
since we are doing it, do we want to pass a context too?

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:466]: to wait() method? yeah probably

Comment (aojea) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:466]: heh https://github.com/golang/go/issues/16620

Comment (aojea) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:500]: I think that is easier to synchronize using only channels.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 2]
```diff
@@ -644,7 +686,13 @@ func (c *Cacher) GetList(ctx context.Context, key string, opts storage.ListOptio
 	trace := utiltrace.New("cacher list", utiltrace.Field{Key: "type", Value: c.objectType.String()})
 	defer trace.LogIfLong(500 * time.Millisecond)
 
-	c.ready.wait()
+	select {
+	case <-c.readyCh:
+	case <-ctx.Done():
+		return ctx.Err()
```
Comment (aojea) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:692]: the context cancellation is racy, if `isReady` but `context is cancelled`

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 3]
```diff
@@ -426,6 +426,36 @@ func (c *Cacher) startCaching(stopChannel <-chan struct{}) {
...
+	default:
+		return false
+	}
+}
+
+// set cacher to ready
+func (c *Cacher) unsetReady() {
+	select {
+	case <-c.readyCh:
+		c.readyCh = make(chan struct{})
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:443]: hmm - this isn't thread-safe, right?

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:443]: I'm personally not against synchronizing using channels, but I think I would rather:

- leave the concept `ready` type as it was
- leave the locking inside it (as we need it to make things like unsetReady thread-safe)
- switch the internal mechanism of it to use channel

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go (hunk 2 above)
Comment (aojea) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:692]: I'll add a comment, the context can be cancelled here or later in that case, this only handles the case when 
"is waiting for ready but not stopped" and the context is cancelled

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go (hunk 3 above)
Comment (aojea) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:443]: to be honest, I thought that this was thread-safe but now I'm not sure
```
	select {
	case <-c.readyCh:
		c.readyCh = make(chan struct{})
```

I found the concept of ready easy to follow too, but the dependency on stop make the sync.Cond complex ... let me try again

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 4]
```diff
@@ -1426,35 +1434,120 @@ func (c *cacheWatcher) process(ctx context.Context, resourceVersion uint64) {
...
+	readyCh chan struct{}
+	stopCh  chan struct{}
 }
 
+// newReady creates a ready condition variable that is not stopped and not ready
 func newReady() *ready {
-	return &ready{c: sync.NewCond(&sync.RWMutex{})}
+	return &ready{
+		readyCh: make(chan struct{}), // is ready if it does not exist
+		stopCh:  make(chan struct{}), // is stopped if it does not exist
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:1452]: I'm not strongly against it, but I'm wondering if we can' achieve what we want by setting to not-ready instead?

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 5]
```diff
@@ -1426,35 +1434,120 @@ func (c *cacheWatcher) process(ctx context.Context, resourceVersion uint64) {
...
+// - it is stopped and returns an error
+// - context is cancelled and returns error
+func (r *ready) wait(ctx context.Context) error {
+	select {
+	// if it is ready but context is cancelled
+	// context cancellation fails fast.
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+		// readyCh and stopCh must not be closed at the same time
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:1467]: nit: can we avoid nesting here?

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 6]
```diff
@@ -1426,35 +1434,120 @@ func (c *cacheWatcher) process(ctx context.Context, resourceVersion uint64) {
 	}
 }
 
+// ready is a condition variable that represent readiness and
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:1437]: this grew quite a bit - can you maybe move that to a separate file (like pkg/storage/cache/ready.go and ready_test.go)

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 7]
```diff
@@ -1426,35 +1434,120 @@ func (c *cacheWatcher) process(ctx context.Context, resourceVersion uint64) {
...
 func (r *ready) check() bool {
-	rwMutex := r.c.L.(*sync.RWMutex)
-	rwMutex.RLock()
-	defer rwMutex.RUnlock()
-	return r.ok
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	select {
+	// if is stopped, the channel is closed, is not ready
+	case <-r.stopCh:
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:1485]: This isn't needed I think - thanks to locking. I think that:
```
select{
case <-r.stopCh:
case <-r.readyCh
default:
}
```

is enough (as when stopCh is closed, then readyCh is not closed)

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 8]
```diff
@@ -1426,35 +1434,120 @@ func (c *cacheWatcher) process(ctx context.Context, resourceVersion uint64) {
...
+		}
+	}
 }
 
 func (r *ready) set(ok bool) {
-	r.c.L.Lock()
-	defer r.c.L.Unlock()
-	r.ok = ok
-	r.c.Broadcast()
+	// close the readyCh if exist so it no longer blocks on wait
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:1499]: With the code below, I think it would probably be easier to have two methods: set() and unset() [as they no longer share code]

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go (hunk 5 above)
Comment (aojea) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:1467]: is an edge case, but we do need the nesting or it flakes if it is ready and the context is cancelled, 50% returns the error of the context and the other 50% returns the ready

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go (hunk 4 above)
Comment (aojea) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:1452]: if not ready we block on the "readyCh" on wait, so we need to unblock on stop without deleting the readyCh, that will mean that is ready.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go (hunk 7 above)
Comment (aojea) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:1485]: yep

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go (hunk 8 above)
Comment (aojea) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:1499]: yes

Comment (aojea) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:1499]: @wojtek-t if is stopped, and then I set it to ready ... what is the expected behavior , it becomes started and ready or is a noop?

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go (hunk 6 above)
Comment (aojea) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:1437]: ack

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/ready.go [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+
+// wait blocks until:
+// - it is ready and returns nil
+// - it is stopped and returns an error
+// - context is cancelled and returns error
+func (r *ready) wait(ctx context.Context) error {
+	select {
+	// if it is ready but context is cancelled
+	// context cancellation fails fast.
+	case <-ctx.Done():
```
Comment (aojea) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/ready.go:51]: testcase `TestCacheWatcherContextIfReady` 
https://github.com/kubernetes/kubernetes/pull/108414/files#diff-aed72203cf711bc5f518b68da5f6131358a0a79139efcbfa675f991951712168R604

without prioritization the test is flaky

**Comment by aojea** [comment]
@liggitt how is it possible that the job reports ok if there are failing tests?

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/108414/pull-kubernetes-integration/1499064087785508864

**Comment by aojea** [comment]
hmm , this is showing some results, there are some tests that fail because of this `TestImpersonateWithUID`

```

	logging error output: "k8s\x00\n\f\n\x02v1\x12\x06Status\x12\x1f\n\x06\n\x00\x12\x00\x1a\x00\x12\aFailure\x1a\astopped\"\x000\xf4\x03\x1a\x00\"\x00"
 >
I0302 17:10:49.509572  118725 get.go:260] "Starting watch" path="/apis/scheduling.k8s.io/v1/priorityclasses" resourceVersion="10078" labels="" fields="" timeout="6m49s"
E0302 17:10:49.509610  118725 status.go:71] apiserver received an error that is not an metav1.Status: storage.InternalError{Reason:"stopped"}: stopped
I0302 17:10:49.508989  118725 get.go:260] "Starting watch" path="/apis/rbac.authorization.k8s.io/v1/rolebindings" resourceVersion="10579" labels="" fields="" timeout="7m22s"
E0302 17:10:49.509715  118725 status.go:71] apiserver received an error that is not an metav1.Status: storage.InternalError{Reason:"stopped"}: stopped
E0302 17:10:49.509901  118725 reflector.go:138] k8s.io/client-go/informers/factory.go:134: Failed to watch *v1.Endpoints: an error on the server ("unknown") has prevented the request from succeeding (get endpoints)
I0302 17:10:49.509842  118725 httplog.go:131] "HTTP" verb="WATCH" URI="/api/v1/namespaces/kube-system/configmaps?allowWatchBookmarks=true&resourceVersion=9835&timeout=9m45s&timeoutSeconds=585&watch=true" latency="8.341339ms" userAgent="Go-http-client/2.0" audit-ID="73d02aef-1515-45b2-8b88-1e2e1c4d471d" srcIP="127.0.0.1:45958" apf_pl="exempt" apf_fs="exempt" apf_fd="" apf_init_latency="7.35213ms" resp=500 statusStack=<
```

/test pull-kubernetes-integration

**Comment by aojea** [comment]
hmm it passed now, maybe the test passes but the log confuses the parser?
/test pull-kubernetes-integration

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/ready.go [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+		return ctx.Err()
+	default:
+		// readyCh and stopCh must not be closed at the same time
+		select {
+		case <-r.readyCh:
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/ready.go:56]: Hmm - I started wondering if this function is actually thread-safe.
Because unset (in line 106) is actually modifies that.

It seems that your tests do check concurrent calls for set(), but not for unset() ...

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/ready.go (hunk 9 above)
Comment (aojea) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/ready.go:51]: agree, is not completely race free, is just solving the race in case the signal are set before entering the wait(), but is just making the behavior more predictable, without the prio is totally random https://go.dev/play/p/QX7RljvRwyu

Besides it races, this is just short circuiting, if not, the code later has to honor the context

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/ready.go (hunk 10 above)
Comment (aojea) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/ready.go:56]: This channel is the synchronization signal , on `check` we avoid manipulating the state meanwhile we are checking the condition but on `wait` ... I can't see other way to implement this, you can't hold a lock meanwhile you are waiting, sync.Cond AFAIK operates the same, once it Wait it releases the lock and has to check the status again after it receives the .Signal and wakes up, since we use a channel to coordinate , it dictates the status inside the select clause

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/ready.go:56]: sync.Cond (the original approach) didn't have this problem as we were checking the env var from scratch (it internally releases the lock as a last thing before going to idle mode).

The only way I can see is to switch back to the original sync.Cond approach but:
1) adding support for stopCh is easy - we can just have the stop() method that will adjust the value of added stopped field and signal all waiters
2) but I don't see an easy way for honoring context then without starting an additional goroutine for every waiter.

OTOH - we would have to start this goroutine only if the fast path of checking ready is false, so that would generally not fire, so maybe that's actually ok and good enough?

Comment (aojea) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/ready.go:56]: SGTM, let's go simple and solve the stop problem now ... the context was a nice to have but is not real a problem now

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/ready.go [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+	defer r.c.L.Unlock()
+	if r.stopped {
+		return
+	}
+	r.ok = ok
+	r.c.Broadcast()
+}
+
+// start the condition variable and set it as not ready
+func (r *ready) start() {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/ready.go:73]: Can we remove it?

The initial values are the default ones, and it's not used in production code, so I suggest removing.

**Review by wojtek-t — COMMENTED** [review]
Just one comment - other than that LGTM

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/ready.go (hunk 11 above)
Comment (aojea) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/ready.go:73]: sure, I just had it for completeness

**Comment by wojtek-t** [comment]
/lgtm
/approve

Thanks!

**Comment by aojea** [comment]
> Commit	Details	Required	Rerun command
> pull-kubernetes-verify	[1f40287](https://github.com/kubernetes/kubernetes/commit/1f40287dfe61a0aa5819709e39b059bb31df7781)	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/108414/pull-kubernetes-verify/1499343734729347072)	true	`/test pull-kubernetes-verify`

missing space between the header and the import statement, @wojtek-t I need a new lgtm please

**Comment by liggitt** [comment]
a couple questions:
* were the added unit tests run with stress / -race?
* is InternalError consistent with what uncached storage returns when the apiserver is shutting down? does that actually surface a 500 to the user? do we want a 503 instead?

**Comment by aojea** [comment]
> * were the added unit tests run with stress / -race?

yeah, absolutely, was the first thing I did

```
go test k8s.io/apiserver/pkg/storage/cacher -race -c
$ stress ./cacher.test -test.v -test.run Test_newReady*
5s: 810 runs so far, 0 failures
10s: 1634 runs so far, 0 failures
15s: 2454 runs so far, 0 failures
20s: 3281 runs so far, 0 failures
25s: 4102 runs so far, 0 failures
30s: 4925 runs so far, 0 failures
35s: 5748 runs so far, 0 failures
^C
$ stress ./cacher.test -test.v -test.run TestCacheDontAcceptRequestsStopped
5s: 812 runs so far, 0 failures
10s: 1636 runs so far, 0 failures
15s: 2461 runs so far, 0 failures
20s: 3288 runs so far, 0 failures
25s: 4106 runs so far, 0 failures
30s: 4927 runs so far, 0 failures
35s: 5746 runs so far, 0 failures
40s: 6564 runs so far, 0 failures
45s: 7389 runs so far, 0 failures
```

> * is InternalError consistent with what uncached storage returns when the apiserver is shutting down? does that actually surface a 500 to the user? do we want a 503 instead?


initially we were considering the context, so IMHO, 500 sounded good, but now that we are only handling the shutdown process, it seems 503 is a better  fit. WDY?

**Comment by aojea** [comment]
/hold
for the 500 and 503 answer

**Comment by liggitt** [comment]
> initially we were considering the context, so IMHO, 500 sounded good, but now that we are only handling the shutdown process, it seems 503 is a better fit. WDYT?

if it surfaces to the user, an error returned because we're not ready yet or because we're shutting down should be a 503, in my opinion

**Comment by aojea** [comment]
PTAL

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 12]
```diff
@@ -644,7 +648,9 @@ func (c *Cacher) GetList(ctx context.Context, key string, opts storage.ListOptio
 	trace := utiltrace.New("cacher list", utiltrace.Field{Key: "type", Value: c.objectType.String()})
 	defer trace.LogIfLong(500 * time.Millisecond)
 
-	c.ready.wait()
+	if err := c.ready.wait(); err != nil {
+		return errors.NewServiceUnavailable(err.Error())
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:652]: I always lose track of what error has to be returned from the storage layer to trigger a 503 API error... did you verify this gets interpreted/propagated correctly by InterpretGetError/InterpretListError/InterpretWatchError

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:652]: if this bubbles correctly to the API, then lgtm

Comment (aojea) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:652]: I was thinking about it after submitting the PR meanwhile and I was going to test it locally, hold on

Comment (aojea) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:652]: ok, before doing more changes by intuition like I'm doing let me get this clear
Tt seems that we want `storage.IsUnreachable(err)` for Get and GetList so it retries (it is a 500 :shrug: )

https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/apiserver/pkg/storage/errors/storage.go#L31-L32

https://github.com/kubernetes/kubernetes/blob/fbdd0d7b4165bc5a677d45e4dc693e3260297bfa/staging/src/k8s.io/apimachinery/pkg/api/errors/errors.go#L361-L376

and for the watch?

https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/apiserver/pkg/storage/errors/storage.go#L106-L116

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:652]: hrmm... I'm fuzzy on the situations that lead us to return this error, and whether timeout or unavailable is what we want to communicate. @wojtek-t @sttts, do you have input on what is downstream of this that will react to the returned error, and which is more correct?

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 13]
```diff
@@ -1040,7 +1047,9 @@ func filterWithAttrsFunction(key string, p storage.SelectionPredicate) filterWit
 
 // LastSyncResourceVersion returns resource version to which the underlying cache is synced.
 func (c *Cacher) LastSyncResourceVersion() (uint64, error) {
-	c.ready.wait()
+	if err := c.ready.wait(); err != nil {
+		return 0, errors.NewServiceUnavailable(err.Error())
+	}
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:1052]: where is this used? The other call sites look clearer, representing CRUD operation. Not this one.

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:1052]: IIRC it was used historically but now it is only used in tests I think.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go (hunk 12 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:652]: If possible an integration test, e.g. k8s.io/apiserver/pkg/server/genericapiserver_test.go TestGracefulTermination would be good, to see that 50x really reaches the client.

@liggitt I guess in our case this is about clients not going through the load-balancer, but localhost (might be related to missing hairpinning support on some cloud platforms), that get stuck on shutdown of the local apiserver although they should connect to another master host.

File: staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store_test.go [hunk 14]
```diff
@@ -2298,6 +2299,60 @@ func TestStoreWatch(t *testing.T) {
 	}
 }
 
+// regression test for #108414
```
Comment (aojea) [inline staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store_test.go:2302]: @wojtek-t this covers a bit more than the unit tests I've added before

File: staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store_test.go [hunk 15]
```diff
@@ -2298,6 +2299,60 @@ func TestStoreWatch(t *testing.T) {
...
+	wi2, err := registry.WatchPredicate(context.TODO(), matchPodName("foo"), "0")
+	if err == nil {
+		t.Errorf("expected error")
+		wi2.Stop()
+	} else {
+		statusError, ok := err.(*errors.StatusError)
+		if !ok {
+			t.Fatalf("Unexpected error type %T", err)
+		}
+		if statusError.Status().Code != http.StatusServiceUnavailable {
```
Comment (aojea) [inline staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store_test.go:2350]: we still didn't decide on the error code

File: staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store_test.go (hunk 14 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store_test.go:2302]: You're effectively going into details of cacher in the test (e.g. casting to it, stopping it, etc.).

Can you move this test down the stack, and make it a test of cacher (under pkg/storage/cacher)?
This feels like a better place for this test.

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store_test.go:2302]: I'm sorry - I can't see that. Can you point me to it?
[I still don't think this is the right place for having such test]

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go [hunk 16]
```diff
@@ -591,6 +601,68 @@ func TestCacheWatcherStoppedOnDestroy(t *testing.T) {
 
 }
 
+func TestCacheDontAcceptRequestsStopped(t *testing.T) {
```
Comment (aojea) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:604]: ping @wojtek-t

File: staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store_test.go (hunk 14 above)
Comment (aojea) [reply staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store_test.go:2302]: pinged you in a new comment, will remove this test. I've added it in a new commit because I already had doubts :smile:

**Comment by wojtek-t** [comment]
This LGTM modulo the unresolved discussion about error to return.

@sttts @liggitt - can you chime in if this looks ok to you?

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 17]
```diff
@@ -1004,6 +1010,7 @@ func (c *Cacher) isStopped() bool {
 
 // Stop implements the graceful termination.
 func (c *Cacher) Stop() {
+	c.ready.stop()
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:1013]: should this be inside `c.stopLock.Lock()` and after the `if c.stopped` check to avoid calling twice?

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/ready.go [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+		r.c.Wait()
+	}
+	if r.stopped {
+		return fmt.Errorf("apiserver cacher is stopped")
+	}
+	return nil
+}
+
+// TODO: Make check() function more sophisticated, in particular
+// allow it to behave as "waitWithTimeout".
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/ready.go:50]: document the return value is `ready and !stopped`

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/ready.go [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+
+import (
+	"fmt"
+	"sync"
+)
+
+// ready is a condition variable that blocks until is ready if is not stopped
+type ready struct {
+	ok      bool
+	stopped bool // if stopped is true ok must be false
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/ready.go:27]: non-blocking, but having two vars to reason about with a constraint between them seems tricker than a single state var that is one of NotReady, Ready, or Stopped.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/ready.go [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+	return nil
+}
+
+// TODO: Make check() function more sophisticated, in particular
+// allow it to behave as "waitWithTimeout".
+func (r *ready) check() bool {
+	rwMutex := r.c.L.(*sync.RWMutex)
+	rwMutex.RLock()
+	defer rwMutex.RUnlock()
+	if r.stopped {
```
[the remark this answers is not in the document]
Comment (aojea) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/ready.go:55]: when it starts both are false, I think that the tristate comment instead of two vars applies here, I can't remember now why I did this :) , maybe was easier to make it explicitly than thinking about it ...

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/ready.go (hunk 19 above)
Comment (aojea) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/ready.go]: done, tri-state it is  better and more correct

**Comment by wojtek-t** [comment]
Still LGTM modulo the discussion about error code.
@liggitt - does that 503 look ok to you?

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/ready.go [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+	}
+	r.c.Broadcast()
+}
+
+// stop the condition variable and set it as Stopped. This state is irreversible.
+func (r *ready) stop() {
+	r.c.L.Lock()
+	defer r.c.L.Unlock()
+	r.state = Stopped
+	r.c.Broadcast()
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/ready.go:89]: should this no-op if already stopped or does it matter?

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/ready.go [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+func (r *ready) wait() error {
+	r.c.L.Lock()
+	defer r.c.L.Unlock()
+	for r.state == Pending {
+		r.c.Wait()
+	}
+	if r.state == Stopped {
+		return fmt.Errorf("apiserver cacher is stopped")
+	}
+	return nil
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/ready.go:56]: prefer a switch with all expected states and an error default:

```go
switch r.state {
case Ready:
  return nil
case Stopped:
  return fmt.Errorf("apiserver cacher is stopped")
default:
  return fmt.Errorf("unexpected apiserver cache state: %v", r.state)
}
```

Comment (aojea) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/ready.go]: done

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/ready.go (hunk 21 above)
Comment (aojea) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/ready.go]: done

**Comment by wojtek-t** [comment]
Still LGTM - but I guess @liggitt should take another look.
